### PR TITLE
e2e for vip dual stack with security group

### PR DESCRIPTION
**Pull request by zcq98**
# Pull Request

- [x] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Tests

## WHAT

copilot:summary

copilot:poem

## HOW

copilot:walkthrough

### CODE REVIEW

File: test/e2e/vip/e2e_test.go [hunk 1]
```diff
@@ -26,10 +26,86 @@ func makeOvnVip(namespaceName, name, subnet, v4ip, v6ip, vipType string) *apiv1.
...
+	case apiv1.ProtocolIPv4:
+		addIP = fmt.Sprintf("ip addr add %s/24 dev eth0", ip)
+		delIP = fmt.Sprintf("ip addr del %s/24 dev eth0", ip)
+		command = fmt.Sprintf("ping -W 1 -c 1 %s", ip)
+	case apiv1.ProtocolIPv6:
+		addIP = fmt.Sprintf("ip addr add %s/96 dev eth0", ip)
+		delIP = fmt.Sprintf("ip addr del %s/96 dev eth0", ip)
+		command = fmt.Sprintf("ping6 -c 1 %s", ip)
+	}
+	// check aapPod2 ping aapPod1 through vip
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:46]: 注释使用 src dst pod 对照起代码更好理解

File: test/e2e/vip/e2e_test.go [hunk 2]
```diff
@@ -26,10 +26,86 @@ func makeOvnVip(namespaceName, name, subnet, v4ip, v6ip, vipType string) *apiv1.
...
+	framework.ExpectNoError(err)
+	// aapPod1 can not ping aapPod2 vip when ip is deleted
+	stdout, stderr, err = framework.ExecShellInPod(context.Background(), f, namespaceName, srcPod, delIP)
+	framework.Logf("exec %s failed err: %v, stderr: %s, stdout: %s", delIP, err, stderr, stdout)
+	stdout, stderr, err = framework.ExecShellInPod(context.Background(), f, namespaceName, dstPod, command)
+	framework.Logf("exec %s failed err: %v, stderr: %s, stdout: %s", command, err, stderr, stdout)
+	framework.ExpectError(err)
+}
+
+func testVipWithSg(ip, namespaceName, allowPod, denyPod, aapPod, securityGroupName string, f *framework.Framework) {
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:72]:
```suggestion
func testVipWithSG(ip, namespaceName, allowPod, denyPod, aapPod, securityGroupName string, f *framework.Framework) {
```

File: test/e2e/vip/e2e_test.go [hunk 3]
```diff
@@ -26,10 +26,86 @@ func makeOvnVip(namespaceName, name, subnet, v4ip, v6ip, vipType string) *apiv1.
...
+		conditions = fmt.Sprintf("name=ovn.sg.%s.associated.v4", strings.ReplaceAll(securityGroupName, "-", "."))
+	case apiv1.ProtocolIPv6:
+		sgCheck = fmt.Sprintf("ping6 -c 1 %s", ip)
+		conditions = fmt.Sprintf("name=ovn.sg.%s.associated.v6", strings.ReplaceAll(securityGroupName, "-", "."))
+	}
+	// allowPod can ping allowPod with security group
+	stdout, stderr, err := framework.ExecShellInPod(context.Background(), f, namespaceName, allowPod, sgCheck)
+	framework.Logf("exec %s failed err: %v, stderr: %s, stdout: %s", sgCheck, err, stderr, stdout)
+	framework.ExpectNoError(err)
+	// denyPod can not ping allowPod with security group
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:87]: // allowPod can ping allowPod with security group
// denyPod can not ping allowPod with security group

注释感觉不对称，是不是写错了?

File: test/e2e/vip/e2e_test.go [hunk 4]
```diff
@@ -26,10 +26,86 @@ func makeOvnVip(namespaceName, name, subnet, v4ip, v6ip, vipType string) *apiv1.
...
+	stdout, stderr, err = framework.ExecShellInPod(context.Background(), f, namespaceName, denyPod, sgCheck)
+	framework.Logf("exec %s failed err: %v, stderr: %s, stdout: %s", sgCheck, err, stderr, stdout)
+	framework.ExpectError(err)
+
+	ginkgo.By("Checking ovn address_set and lsp port_security")
+	// address_set should have aap IP
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:93]: aap 最好用下全称

File: test/e2e/vip/e2e_test.go [hunk 5]
```diff
@@ -26,10 +26,86 @@ func makeOvnVip(namespaceName, name, subnet, v4ip, v6ip, vipType string) *apiv1.
...
+	framework.ExpectNoError(err)
+	framework.ExpectContainElement(addressSet, ip)
+	// port_security should have aap IP
+	execCmd = fmt.Sprintf("kubectl ko nbctl --format=list --data=bare --no-heading --columns=port_security find logical-switch-port name=%s.%s", aapPod, namespaceName)
+	output, err = exec.Command("bash", "-c", execCmd).CombinedOutput()
+	portSecurity := strings.Split(strings.ReplaceAll(string(output), "\n", ""), " ")
+	framework.ExpectNoError(err)
+	framework.ExpectContainElement(portSecurity, ip)
+	// TODO: Checking allow address pair connectivity with security group
+	// AAP does not work fine with security group in kind
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:106]:
```suggestion
	// AAP does not work fine with security group in kind test env for now
```

File: test/e2e/vip/e2e_test.go [hunk 6]
```diff
@@ -123,78 +199,65 @@ var _ = framework.Describe("[group:vip]", func() {
...
-			framework.ExpectEqual(expectVirtualParents, virtualParents)
-			// other pods can communicate with the aap pod through vip
+		}
+		virtualParents := strings.Split(options["virtual-parents"], ",")
+		sort.Strings(virtualParents)
+		expectVirtualParents := []string{fmt.Sprintf("%s.%s", aapPodName1, namespaceName), fmt.Sprintf("%s.%s", aapPodName2, namespaceName)}
+		sort.Strings(expectVirtualParents)
+		framework.ExpectEqual(expectVirtualParents, virtualParents)
+
+		ginkgo.By("Test aap address connectivity")
```
Comment (bobz965) [inline test/e2e/vip/e2e_test.go:245]: 命名可以缩写，注释和 log 尽量用 aap 全称

**Review by bobz965 — APPROVED** [review]

File: test/e2e/vip/e2e_test.go (hunk 6 above)
Comment (zcq98) [reply test/e2e/vip/e2e_test.go:245]: 已修复

File: test/e2e/vip/e2e_test.go (hunk 1 above)
Comment (zcq98) [reply test/e2e/vip/e2e_test.go:46]: 已修复

File: test/e2e/vip/e2e_test.go (hunk 2 above)
Comment (zcq98) [reply test/e2e/vip/e2e_test.go:72]: 已修复

File: test/e2e/vip/e2e_test.go (hunk 3 above)
Comment (zcq98) [reply test/e2e/vip/e2e_test.go:87]: 已修复

File: test/e2e/vip/e2e_test.go (hunk 5 above)
Comment (zcq98) [reply test/e2e/vip/e2e_test.go:106]: 已修复

**Pull request merged**